File: packages/mirascript/src/helpers/convert/to-format.ts
import type { VmAny } from '../../vm/index.js';
import { isFinite } from '../utils.js';
import { toString } from './to-string.js';

/** 渲染数字 */
function formatNumber(finite: number): string {
    if (finite === 0) return '0';
    const s = finite.toString();
    let ps;
    const abs = Math.abs(finite);
    if (abs >= 1000 || abs < 0.001) {
        const ps1 = finite.toExponential();
        const ps2 = finite.toExponential(5);
        ps = ps1.length < ps2.length ? ps1 : ps2;
    } else {
        ps = finite.toPrecision(6);
    }
    return ps.length < s.length ? ps : s;
}

/** 格式化为 string */
export function toFormat(value: VmAny, format: string | null | undefined): string {
    const f = format == null ? '' : format.trim();

    if (typeof value == 'number') {
        if (!isFinite(value)) {
            return toString(value, undefined as never);
        }
        if (/^\.\d+$/.test(f)) {
            let digits = Math.trunc(Number(f.slice(1)));
            if (!(digits <= 100)) digits = 100;
            return value.toFixed(digits);
        } else {
            return formatNumber(value);
        }
    }

    return toString(value, undefined as never);
}

File: crates/python/mirascript/helpers/convert/to_format.py
import math
import re

from .to_string import toString


def formatNumber(num):
    if num == 0:
        return "0"

    s = toString(num)

    ps = ""
    absVal = abs(num)
    if absVal >= 1000 or absVal < 0.001:
        ps1 = format(num, f".0e")
        ps2 = format(num, f".5e")
        ps = ps1 if len(ps1) < len(ps2) else ps2

    else:
        ps = format(float(num), f".6")

    return ps if len(ps) < len(s) else s


def toFormat(val, fmt: "str | None" = None) -> str:
    fmt = fmt.strip() if fmt is not None else ""

    if isinstance(val, bool):
        return toString(val)

    if isinstance(val, (int, float)):
        if not math.isfinite(val):
            return toString(val)
        r = re.match(r"^\.\d+$", fmt)
        if r:
            ff = float(fmt[1:])
            if ff < 0:
                ff = 0
            elif ff > 100:
                ff = 100
            return f"{val:.{math.trunc(ff)}f}"
        else:
            return formatNumber(val)
    return toString(val)
